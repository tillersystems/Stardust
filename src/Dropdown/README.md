# Dropdown

### Usage

```jsx
import { Dropdown } from '@tillersystems/stardust';
```

<!-- STORY -->

### Description

A Dropdown displays content through its children prop that must be components wrapping text.  
The trigger is a button displaying text provided by the prop `title`.

### Properties

| Name                   | Required |    Type    | DefaultValue |                                                             Description                                                              |
| ---------------------- | :------: | :--------: | :----------: | :----------------------------------------------------------------------------------------------------------------------------------: |
| `children`             |    \*    |   `node`   |              |                         Anything that can be rendered: numbers, strings, elements or an array (or fragment)                          |
| `className`            |    -     |  `string`  |    `null`    |                                       Adds a text aside in the select next the selected value                                        |
| `contentRef`           |    -     | `function` | `undefined`  |                                                   Callback ref of content element                                                    |
| `displayMenu`          |    -     |   `bool`   | `undefined`  | If the dropdown is open or not. If it is in a controlled state, this prop should be passed, otherwise it will rely on internal state |
| `headerStyle`          |    -     |  `Object`  |    `null`    |                                                      Style for header component                                                      |
| `itemCss`              |    -     |  `array`   |    `null`    |                       CSS provided to each item of the dropdown. Must use `css` method from styled-components                        |
| `modifiers`            |    -     |  `object`  |    `null`    |    Customize popper behaviour. Plugins to alter the behaviour of the popper. See https://popper.js.org/popper-documentation.html     |
| `noResultLabel`        |    -     |   `node`   |    `null`    |                                               Label to display when no result is found                                               |
| `onToggle`             |    -     | `function` |  `() => {}`  |                                               Callback called when Dropdown is toggled                                               |
| `searchable`           |    -     |   `bool`   |   `false`    |                                                  Whether the dropdown is searchable                                                  |
| `searchBarPlaceholder` |    -     |  `string`  |     `''`     |                                                     SearchBar input placeholder                                                      |
| `title`                |    \*    |   `node`   |              |                                                            Dropdown title                                                            |
| `usePortal`            |    -     |   `bool`   |   `false`    |                                                   Display the content on a portal                                                    |
| `...listProps`         |    -     |  `object`  |   `false`    |                               Dropdown accept `OptionsList` props to render a list instead of children                               |

### Example

```jsx
import { Dropdown } from '@tillersystems/stardust';

render() {
  // using children
  return (
    <Dropdown title="Filter by">
      <div>
        children
      </div>
    </Dropdown>
  );

  // passing OptionsList props without children
  return (
    <Dropdown
      title="Filter by"
      // OptionsList props below...
      allowMultiple
      searchMethod={searchMethod}
      onChange={onChange}
      options={options}
      searchBarPlaceholder={SearchBarPlaceholder}
      noResultLabel={NoResultLabel}
      values={state.selectedStores}
      />
  );
}
```
