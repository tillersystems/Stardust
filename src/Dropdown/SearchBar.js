import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import TextInput from '../Input/TextInput';

const { func, string } = PropTypes;

/**
 * SearchBar
 *
 * This component is in charge of displaying
 * a search bar
 *
 * @param {string} children // Anything that can be rendered: numbers, strings, elements or an array (or fragment).
 * @param {string} className // Add a text aside in the select next the selected value.
 * @param {function} onChange // Callback function called when text input change.
 * @param {string} placeHolder // Input placeHolder.
 * @param {string} value // Input value.
 *
 * @return {jsx}
 */

class SearchBar extends PureComponent {
  /** Prop types. */
  static propTypes = {
    className: string,
    onChange: func.isRequired,
    placeHolder: string,
    value: string,
  };

  /** Default props. */
  static defaultProps = {
    className: '',
    placeHolder: '',
    value: '',
  };

  /**
   * handleChange
   */
  handleChange = searchTerm => {
    const { onChange } = this.props;

    onChange && onChange(searchTerm);
  };

  /**
   * Renders the component.
   *
   * @return {jsx}
   */
  render() {
    const { className, placeHolder, value } = this.props;

    return (
      <div className={className}>
        <TextInput
          ghost
          data-test="search-input"
          value={value}
          onChange={this.handleChange}
          placeHolder={placeHolder}
          fluid
          search
        />
      </div>
    );
  }
}

export default styled(SearchBar)`
  width: 100%;
  border-bottom: 1px solid ${({ theme: { palette } }) => palette.lightGrey};
`;
