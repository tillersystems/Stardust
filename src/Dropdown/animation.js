/**
 * Define animation variants for modal animation
 * https://www.framer.com/api/motion/
 */
export const animationVariants = {
  visible: {
    opacity: 1,
    scaleY: 1,
    originX: '50%',
    originY: '0%',
    transition: {
      ease: 'anticipate',
      duration: 0.15,
    },
  },
  hidden: {
    opacity: 0,
    scaleY: 0,
    originX: '50%',
    originY: '0%',
    transition: {
      duration: 0.15,
    },
  },
};
