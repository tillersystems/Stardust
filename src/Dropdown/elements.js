import styled from 'styled-components';

export const Header = styled.button`
  display: flex;
  align-items: center;

  height: 100%;
  width: 100%;
  position: relative;

  padding: 0 1.2rem;

  cursor: pointer;
  text-align: left;
  font-size: ${({ theme: { fonts } }) => fonts.size.medium};

  border: 1px solid ${({ theme: { palette } }) => palette.lightGrey};
  border-radius: ${({ theme: { dimensions } }) => dimensions.radius};

  background: linear-gradient(
    180deg,
    ${({ theme: { palette } }) => palette.white} 0%,
    ${({ theme: { palette } }) => palette.paleGrey} 100%
  );
  color: ${({ theme: { palette } }) => palette.spaceGrey};

  &::after {
    content: '';

    position: absolute;
    right: 1.4rem;
    top: 37%;

    width: 0;
    height: 0;

    border: 0 solid transparent;
    border-right-width: 3px;
    border-left-width: 3px;
    border-bottom: 3px solid ${({ theme: { palette } }) => palette.spaceGrey};
  }
  &::before {
    content: '';

    position: absolute;
    right: 1.4rem;
    bottom: 33%;

    width: 0;
    height: 0;

    border: 0 solid transparent;
    border-left-width: 3px;
    border-right-width: 3px;
    border-top: 3px solid ${({ theme: { palette } }) => palette.spaceGrey};
  }
`;

export const HeaderContent = styled.div`
  align-items: center;
  display: flex;
  height: 2.2rem;
  padding-right: 2rem;
`;

export const Menu = styled.ul`
  width: 100%;
  max-height: 28rem;

  list-style: none;
  overflow-y: auto;

  color: ${({ theme: { palette } }) => palette.spaceGrey};

  transform-origin: 50% 0%;
`;

export const MenuItem = styled.li`
  display: flex;
  align-items: center;
  cursor: pointer;
`;

export const SearchInputContainer = styled(MenuItem)`
  padding: 0;
  &:first-of-type {
    padding: 0;
  }
`;
