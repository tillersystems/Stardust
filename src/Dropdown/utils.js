/**
 * Return the offset position of a DOM element
 * @param {*} node DOM element
 *
 * @returns {object} offset position
 */
export const offset = node => {
  const rect = node.getBoundingClientRect();
  const scrollLeft = window.pageXOffset || document.documentElement.scrollLeft;
  const scrollTop = window.pageYOffset || document.documentElement.scrollTop;

  return { top: rect.top + scrollTop, left: rect.left + scrollLeft };
};
