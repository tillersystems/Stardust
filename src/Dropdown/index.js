/* eslint-disable react/require-default-props */

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Popover from '../Popover';
import { Header, HeaderContent } from './elements';
import { animationVariants } from './animation';
import { offset } from './utils';
import OptionsList from '../OptionsList';

/**
 * A Dropdown displays content through its children prop that must be components wrapping text.
 * The trigger is a button displaying text provided by the prop `title`
 *
 * See README.md and storybook for more documentation
 *
 * @return {jsx}
 */

class Dropdown extends PureComponent {
  /** Internal state. */
  state = {
    displayMenu: false,
    searchKeyword: '',
    portalPosition: { left: '0px', top: '0px' },
  };

  /** preserve the initial state in a new object. */
  baseState = this.state;

  constructor(props) {
    super(props);
    this.dropdown = React.createRef();
  }

  componentDidMount() {
    // listen for resize and mousewheel to handle portal position
    setTimeout(this.updatePosition.bind(this), 0);
    window.addEventListener('resize', this.updatePosition.bind(this));
    document.body.addEventListener('mousewheel', this.updatePosition.bind(this));
  }

  componentWillUnmount() {
    // will remove resize and mousewheel event on dismount
    window.removeEventListener('resize', this.updatePosition.bind(this));
    document.body.removeEventListener('mousewheel', this.updatePosition.bind(this));
  }

  /**
   * Handle Search
   */
  handleSearch = searchTerm => {
    this.setState({ ...this.state, searchKeyword: searchTerm });
  };

  /**
   * Returns true if requested prop exists
   *
   * @param {string} prop - requested property name
   *
   * @return {boolean}
   */
  isControlled = prop => {
    const props = this.props;

    return props.hasOwnProperty(prop);
  };

  /**
   * Returns the value a property from props if prop exists and from state otherwise
   *
   * @param {string} key - requested property name
   *
   * @return {*} value of the property
   */
  getControllableValue = key => {
    const props = this.props;
    const state = this.state;

    return this.isControlled(key) ? props[key] : state[key];
  };

  /**
   * Closes the select on outside click and triggers callback of parent
   *
   */
  onClickOutside = event => {
    const { onToggle } = this.props;

    if (this.isControlled('displayMenu')) {
      onToggle && onToggle(event);
    } else {
      this.setState(
        {
          displayMenu: false,
        },
        () => {
          onToggle && onToggle(false);
        },
      );
    }
  };

  /**
   * Reset Search Input
   * Clear search input value
   *
   */
  resetSearchInput = () => {
    const { portalPosition } = this.state;
    this.setState({ ...this.baseState, portalPosition });
  };

  /**
   * Toggle Menu display and clear search input.
   * Triggers callback if parent provided it
   *
   */
  toggleMenu = () => {
    const displayMenu = this.getControllableValue('displayMenu');
    const { onToggle } = this.props;

    this.updatePosition();

    if (this.isControlled('displayMenu')) {
      onToggle && onToggle(event);
    } else {
      this.setState(
        prevState => ({
          displayMenu: !prevState.displayMenu,
        }),
        () => {
          // Clear input value if menu is open
          if (displayMenu) {
            this.resetSearchInput();
          }
          onToggle && onToggle(!displayMenu);
        },
      );
    }
  };

  /**
   * Triggered when ref to the trigger has been set.
   * Allows us to retrieve the width of the trigger to set it to the content wrapper
   *
   */
  triggerRef = ref => {
    const contentWidth = ref && ref.offsetWidth;

    this.setState({ contentWidth: `${contentWidth}px` });
  };

  /**
   * Save current dropdown position in state for portal
   */
  updatePosition() {
    if (this.dropdown.current) {
      this.setState({ portalPosition: offset(this.dropdown.current) });
    }
  }

  /**
   * Renders the component.
   *
   * @return {jsx}
   */
  render() {
    const {
      children,
      className,
      contentRef,
      headerStyle,
      modifiers,
      title,
      usePortal,
      ...listProps
    } = this.props;
    const { contentWidth, displayMenu: displayMenuState, portalPosition } = this.state;
    const { displayMenu: displayMenuProp } = this.props;
    const displayMenu = this.isControlled('displayMenu') ? displayMenuProp : displayMenuState;

    return (
      <div ref={this.dropdown} className={className} data-testid="dropdown">
        <Popover
<<<<<<< HEAD
          content={
            <Menu>
              {searchable && (
                <SearchInputContainer>
                  <SearchBar
                    placeholder={searchBarPlaceholder}
                    onChange={this.handleSearch}
                    value={searchKeyword}
                  />
                </SearchInputContainer>
              )}

              {searchable &&
                (FilteredItems.length !== 0
                  ? FilteredItems.map(child => (
                      <MenuItem
                        key={child.key}
                        role="menuitem"
                        css={itemCss}
                        searchable={searchable}
                      >
                        {child}
                      </MenuItem>
                    ))
                  : noResultLabel && (
                      <MenuItem role="menuitem" css={itemCss}>
                        {noResultLabel}
                      </MenuItem>
                    ))}
              {!searchable &&
                Children.map(children, child => (
                  <MenuItem key={child} role="menuitem" css={itemCss}>
                    {child}
                  </MenuItem>
                ))}
            </Menu>
          }
=======
          animationVariants={animationVariants}
          content={children || <OptionsList {...listProps} />}
>>>>>>> f4bae39... feat(Dropdown): can render an OptionsList instead of children
          contentRef={contentRef}
          contentWrapperStyle={{ marginTop: '4rem' }}
          isOpen={displayMenu}
          modifiers={modifiers}
          onClickOutside={this.onClickOutside}
          portalPosition={portalPosition}
          triggerRef={this.triggerRef}
          usePortal={usePortal}
          width={contentWidth}
        >
          <Header
            onClick={this.toggleMenu}
            aria-haspopup="true"
            aria-expanded={displayMenu}
            data-testid="dropdown-button"
            style={headerStyle}
          >
            <HeaderContent>{title}</HeaderContent>
          </Header>
        </Popover>
      </div>
    );
  }
}

const { bool, func, object, node, string } = PropTypes;

/**
 *
 * PropTypes validation
 */
Dropdown.propTypes = {
  /**
   * Anything that can be rendered: numbers, strings, elements or an array (or fragment)
   */
  children: node,

  /**
   * ClassName needed by styled components
   */
  className: string,

  /**
   * Callback ref of content element
   */
  contentRef: func,

  /**
   * If the dropdown is open or not. If it is in a controlled state, this prop should be passed
   * otherwise it will rely on internal state
   */
  // eslint-disable-next-line react/no-unused-prop-types
  displayMenu: bool,

  /**
   * Style for header component
   */
  headerStyle: object,

  /**
   * CSS height of the component
   */
  // eslint-disable-next-line react/no-unused-prop-types
  height: string,

  /**
   * OptionsList props, if innerComponent is null, we will render the children
   */
  innerComponent: object,

  /**
   * Customize popper behaviour. Plugins to alter the behaviour of the popper. See https://popper.js.org/popper-documentation.html
   */
  modifiers: object,

  /**
   * Callback called when Dropdown is toggled
   */
  onToggle: func,

  /**
   * Dropdown title
   */
  title: node.isRequired,

  /**
   * Display the content on a portal
   */
  usePortal: bool,
};

/**
 * Default props
 */
Dropdown.defaultProps = {
  children: null,
  className: '',
  headerStyle: null,
  height: '3.8rem',
  innerComponent: null,
  modifiers: null,
  onToggle: () => {},
  usePortal: false,
};

export default styled(Dropdown)`
  position: relative;
  user-select: none;
  height: ${({ height }) => height || '3.8rem'};
`;
