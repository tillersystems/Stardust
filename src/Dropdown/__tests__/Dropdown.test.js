import React from 'react';
import { fireEvent, waitForElementToBeRemoved } from '@testing-library/react';
import { css } from 'styled-components';

import Dropdown from '..';

jest.mock('popper.js');

describe('<Dropdown />', () => {
  test('should render without a problem', () => {
    const props = { title: 'title' };
    const { container } = render(
      <Dropdown {...props}>
        <div>Item1</div>
        <div>Item2</div>
        <div>Item3</div>
      </Dropdown>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
  test('should render without a problem using portal', () => {
    const props = { title: 'title' };
    const { getByTestId } = render(
      <Dropdown usePortal {...props}>
        <div>Item1</div>
        <div>Item2</div>
        <div>Item3</div>
      </Dropdown>,
    );

    const node = getByTestId('positioned-portal');
    expect(node).toBeInTheDocument();
  });

  test('should toggle the dropdown', async () => {
    const props = { title: 'title' };
    const { container, queryAllByText } = render(
      <Dropdown {...props}>
        <div>Item1</div>
        <div>Item2</div>
        <div>Item3</div>
      </Dropdown>,
    );
    const button = container.querySelector('button');

    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(button).toHaveAttribute('aria-haspopup', 'true');

    // Open Dropdown
    fireEvent.click(button);

    expect(button).toHaveAttribute('aria-expanded', 'true');
    expect(button).toHaveAttribute('aria-haspopup', 'true');

    const ItemNode = queryAllByText(/Item/);

    expect(ItemNode).toHaveLength(3);
    expect(ItemNode[0]).toBeInTheDocument();
    expect(ItemNode[1]).toBeInTheDocument();
    expect(ItemNode[2]).toBeInTheDocument();

    // Close Dropdown
    fireEvent.click(button);

    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(button).toHaveAttribute('aria-haspopup', 'true');

    const ItemNodeRemoved = await waitForElementToBeRemoved(() => queryAllByText(/Item/));
    expect(ItemNodeRemoved[0]).toBeUndefined();
    expect(ItemNodeRemoved[1]).toBeUndefined();
    expect(ItemNodeRemoved[2]).toBeUndefined();
  });

  test('should filter items', () => {
    const props = { title: 'title', searchable: true, searchBarPlaceholder: 'search' };
    const { container, queryAllByText, getByText, getByPlaceholderText } = render(
      <Dropdown {...props}>
        <div>Item1</div>
        <div>Item2</div>
        <div>Item3</div>
      </Dropdown>,
    );

    const button = container.querySelector('button');

    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(button).toHaveAttribute('aria-haspopup', 'true');

    // Open Dropdown
    fireEvent.click(button);

    expect(button).toHaveAttribute('aria-expanded', 'true');
    expect(button).toHaveAttribute('aria-haspopup', 'true');

    // Type Item2 in search input
    fireEvent.change(getByPlaceholderText('search'), { target: { value: 'Item2' } });

    const ItemNode = queryAllByText(/Item/);
    const Item2Node = getByText('Item2');

    expect(ItemNode).toHaveLength(1);
    expect(Item2Node).toBeInTheDocument();
  });

  test('should return Not Found if not item found', () => {
    const props = {
      title: 'title',
      searchable: true,
      searchBarPlaceholder: 'search',
      noResultLabel: 'Not Found',
    };
    const { container, queryAllByText, getByText, getByPlaceholderText } = render(
      <Dropdown {...props}>
        <div>Item1</div>
        <div>Item2</div>
        <div>Item3</div>
      </Dropdown>,
    );

    const button = container.querySelector('button');

    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(button).toHaveAttribute('aria-haspopup', 'true');

    // Open Dropdown
    fireEvent.click(button);

    expect(button).toHaveAttribute('aria-expanded', 'true');
    expect(button).toHaveAttribute('aria-haspopup', 'true');

    const SearchInput = getByPlaceholderText(props.searchBarPlaceholder);

    // Type Item2 in search input
    fireEvent.change(SearchInput, { target: { value: 'qwerty' } });

    const NotFoundNode = getByText(props.noResultLabel);
    const ItemNode = queryAllByText(/Item/);

    expect(NotFoundNode).toBeInTheDocument();
    expect(ItemNode[0]).toBeUndefined();
    expect(ItemNode[1]).toBeUndefined();
    expect(ItemNode[2]).toBeUndefined();
  });

  test('should have expected custom style on each item', () => {
    const props = {
      title: 'title',
      itemCss: css`
        padding: 0.9rem 1.2rem;
        &:first-child {
          padding-top: 1.8rem;
        }
        &:last-child {
          padding-bottom: 1.8rem;
        }
      `,
    };
    const { container, getAllByRole } = render(
      <Dropdown {...props}>
        <div>Item1</div>
        <div>Item2</div>
        <div>Item3</div>
      </Dropdown>,
    );
    const button = container.querySelector('button');
    fireEvent.click(button);

    const [itemNode] = getAllByRole('menuitem');

    expect(itemNode).toHaveStyleRule('padding', '0.9rem 1.2rem');
  });
});
