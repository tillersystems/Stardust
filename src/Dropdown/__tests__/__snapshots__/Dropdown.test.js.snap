// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should render without a problem 1`] = `
.c1 {
  display: inline-block;
  height: inherit;
  width: inherit;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  width: 100%;
  position: relative;
  padding: 0 1.6rem;
  cursor: pointer;
  text-align: left;
  font-size: 1.4rem;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 0.4rem;
  background: linear-gradient( 180deg,hsl(0,100%,100%) 0%,hsl(240,33%,99%) 100% );
  color: hsl(207,13%,45%);
}

.c2::after {
  content: '';
  position: absolute;
  right: 1.4rem;
  top: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-right-width: 3px;
  border-left-width: 3px;
  border-bottom: 3px solid hsl(207,13%,45%);
}

.c2::before {
  content: '';
  position: absolute;
  right: 1.4rem;
  bottom: 33%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-left-width: 3px;
  border-right-width: 3px;
  border-top: 3px solid hsl(207,13%,45%);
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.2rem;
  padding-right: 2rem;
}

.c0 {
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 3.8rem;
}

<div
  class="c0"
  data-testid="dropdown"
>
  <span
    class="c1"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="c2"
      data-testid="dropdown-button"
    >
      <div
        class="c3"
      >
        title
      </div>
    </button>
  </span>
</div>
`;
